op-service/sources/prefetching_eth_client_test.go

```diff
@@ -0,0 +1,122 @@
+package sources
+
+import (
+	"context"
+	"math/rand"
+	"strconv"
+	"sync"
+	"testing"
+	"time"
+
+	"github.com/ethereum-optimism/optimism/op-service/testutils"
+	"github.com/ethereum/go-ethereum/core/types"
+	"github.com/stretchr/testify/require"
+)
+
+// Define a test type for each method
+type testFunction func(t *testing.T, ctx context.Context, client *PrefetchingEthClient, mock *testutils.MockEthClient)
+
+// Define test cases
+var testCases = []struct {
+	name              string
+	prefetchingRanges []uint64
+	testFunc          testFunction
+}{
+	{
+		name:              "InfoByNumber",
+		prefetchingRanges: []uint64{0, 1, 5},
+		testFunc: func(t *testing.T, ctx context.Context, client *PrefetchingEthClient, mock *testutils.MockEthClient) {
+			randomness := rand.New(rand.NewSource(123))
+
+			// set up a random block to get from the client
+			block, _ := randomRpcBlockAndReceipts(randomness, 2)
+			rhdr := block.rpcHeader
+			expectedTxs := block.Transactions
+			expectedInfo, err := rhdr.Info(true, false)
+			require.NoError(t, err)
+			mock.ExpectInfoAndTxsByNumber(uint64(rhdr.Number), expectedInfo, expectedTxs, nil)
+
+			// also set up a window of random blocks and receipts to prefetch
+			windowEnd := (uint64(rhdr.Number) + client.PrefetchingRange)
+			for i := uint64(rhdr.Number) + 1; i <= windowEnd; i++ {
+				// set up different info per iteration
+				fillerBlock, fillerReceipts := randomRpcBlockAndReceipts(randomness, 2)
+				fillerInfo, err := fillerBlock.rpcHeader.Info(true, false)
+				require.NoError(t, err)
+				mock.ExpectInfoAndTxsByNumber(i, fillerInfo, fillerBlock.Transactions, nil)
+				mock.ExpectFetchReceipts(fillerBlock.Hash, fillerInfo, fillerReceipts, nil)
+			}
+
+			info, txs, err := client.InfoAndTxsByNumber(ctx, uint64(rhdr.Number))
+			require.NoError(t, err)
+			require.Equal(t, info, expectedInfo)
+			require.Equal(t, txs, types.Transactions(expectedTxs))
+		},
+	},
+}
+
+// runTest runs a given test function with a specific prefetching range.
+func runTest(t *testing.T, name string, testFunc testFunction, prefetchingRange uint64) {
+	ctx := context.Background()
+	mockEthClient := new(testutils.MockEthClient)
+	client, err := NewPrefetchingEthClient(mockEthClient, prefetchingRange, 30*time.Second)
```

Comment:
we should also add `defer client.Close()` here to release the resources when the test is done.